src/manager.c

```diff
@@ -21,194 +34,174 @@ get_env_str(const char* env_name)
- * Utility functions
- ******************************************************************************/
-
-void print_full_cap(uintcap_t cap) {
-    uint32_t words[4];  // Hack to demonstrate! In real code, be more careful about sizes, etc.
-    printf("0x%d", cheri_tag_get(cap) ? 1 : 0);
-    memcpy(words, &cap, sizeof(cap));
-    for (int i = 3; i >= 0; i--) {
-        printf("_%08x", words[i]);
-    }
-    printf("\n");
-}
-
-/* Setup required capabilities on the heap to jump from within compartments via
- * a context switch
- *
- * For each function to be intercepted, we define the following:
- * redirect_func function to be executed at a higher privilege level
- * TODO I think the below three are common and can be lifted
- * intercept_ddc ddc to be installed for the transition
- * intercept_pcc
- *      higher privileged pcc pointing to the transition support function
- * sealed_redirect_cap
- *      sealed capability pointing to the consecutive intercept capabilities;
- *      this is the only component visible to the compartments
- */
-void
-setup_intercepts()
-{
-    for (size_t i = 0; i < sizeof(to_intercept_funcs) / sizeof(to_intercept_funcs[0]); ++i)
+    for (size_t i = 0; i < comps_count; ++i)
     {
-        comp_intercept_funcs[i].func_name = to_intercept_funcs[i].func_name;
-        comp_intercept_funcs[i].redirect_func = to_intercept_funcs[i].redirect_func;
-        comp_intercept_funcs[i].intercept_ddc = manager_ddc;
-        comp_intercept_funcs[i].intercept_pcc =
-            cheri_address_set(cheri_pcc_get(), (uintptr_t) intercept_wrapper);
-        void* __capability sealed_redirect_cap =
-            cheri_address_set(manager_ddc, (uintptr_t) &comp_intercept_funcs[i].intercept_ddc);
-        asm("SEAL %[cap], %[cap], lpb\n\t"
-                : [cap]"+C"(sealed_redirect_cap)
-                : /**/ );
-        comp_intercept_funcs[i].redirect_cap = sealed_redirect_cap;
+        if (comps[i]->comp->id == to_find->id)
+        {
+            return comps[i];
+        }
     }
-    comp_return_caps[0] = manager_ddc; // TODO does this need to be sealed?
-    comp_return_caps[1] = cheri_address_set(cheri_pcc_get(), (uintptr_t) comp_exec_out);
+    assert(false && "Couldn't find requested compartment.");
 }
 
 struct Compartment*
-manager_find_compartment_by_addr(void* ptr)
+manager_find_compartment_by_addr(void* addr)
 {
-    return loaded_comp; // TODO
+    size_t i;
+    for (i = 0; i < comps_count; ++i)
+    {
+        if (comps[i]->comp->base <= (uintptr_t) addr &&
+            comps[i]->comp->base + comps[i]->comp->size > (uintptr_t) addr)
+        {
+            break;
+        }
+    }
+    assert(i != comps_count);
+    return comps[i]->comp;
 }
 
 struct Compartment*
 manager_find_compartment_by_ddc(void* __capability ddc)
 {
-    return loaded_comp; // TODO
+    size_t i;
+    for (i = 0; i < comps_count; ++i)
+    {
+        if (comps[i]->comp->ddc == ddc)
+        {
+            break;
```

Comment:
I think it would be clearer to `return` here, then the later `assert` can more easily become `err` or similar.